### ISSUE TITLE
bug: needle_retrieval never feeds attention to the cache, so attention-based policies are scored as all-zero (F13 regression)

### ISSUE THREAD

**Issue by umran666** [umran666/UniqKache#40]
## Summary

`needle_retrieval` in `src/uniqkache/metrics/quality.py` never passes `return_attention=True` to the model and never calls `_absorb_attention`, so attention-consuming policies are scored on a cache whose `cum_attention` stayed all-zero for the entire evaluation. For those policies the score describes a policy that was never run.

## Affected component

- `src/uniqkache/metrics/quality.py` — `needle_retrieval` (prefill at line 337, decode steps at line 347)
- `src/uniqkache/bench/runner.py` — needle quality path (lines 441–456)
- CLI-reachable via `--quality-metric needle_retrieval` (`src/uniqkache/bench/cli.py` lines 151–170)

## Background

`perplexity` was fixed for exactly this hazard — its code (quality.py lines 168–198) states:

```python
# An attention-based policy must be fed attention here, exactly as it is
# during generation. Without this, `cum_attention` stays all-zero for the
# whole quality pass, and every attention-based policy silently scores every
# token equally -- which means the quality number describes a policy that
# was never actually run. See docs/research.md, F13.
record_attention = bool(
    cache is not None and cache.policy is not None and cache.policy.uses_attention
)
...
logits, weights = model.forward(chunk, cache=cache, start_pos=start, return_attention=record_attention)
_absorb_attention(cache, weights, num_queries=end - start)
```

`needle_retrieval` (quality.py 334–354) does neither:

```python
if cache is not None:
    cache.reset()
    logits, _ = model.forward(context, cache=cache, start_pos=0)      # no return_attention
...
    step_logits, _ = model.forward(generated[-1], cache=cache, start_pos=position)  # no return_attention
```

`_absorb_attention`'s own docstring (quality.py 47–64) calls this an F13 recurrence: "If it did not, the quality number would describe a policy that was never run — which is precisely what happened before this existed."

## Observed vs expected

**Observed:** With `--quality-metric needle_retrieval`, the policies `attention_based`, `token_importance`, `adaptive`, and `lru` (all `uses_attention=True`) consume `cum_attention` that is all zeros. Their scoring degenerates to the base-class stable tie-break, and the needle score reflects a cache that no attention-based policy actually ran.

**Expected:** The quality pass feeds attention into the cache exactly like generation does (mirroring `GenerationEngine.generate`, generation.py lines 199–203 and 225–230), so an attention-based policy's needle score describes the policy that ran.

## Steps to reproduce

```bash
python -m uniqkache.bench --model synthetic:tiny \
  --policy attention_based --context-length 256 --keep-ratio 0.25 \
  --quality-metric needle_retrieval --needle-length 6
```

(All flags confirmed live in `bench/cli.py` 151–170. In the `needle_retrieval` call, `cache.policy.uses_attention` is True, yet no attention is ever recorded — grep the needles path: no `return_attention`, no `_absorb_attention`.)

## Suggested fix

Mirror `perplexity`'s pattern in `needle_retrieval`:

1. After `cache.reset()`, compute `record_attention = bool(cache is not None and cache.policy is not None and cache.policy.uses_attention)`.
2. Prefill: `logits, weights = model.forward(context, cache=cache, start_pos=0, return_attention=record_attention)`, then `_absorb_attention(cache, weights, num_queries=context.shape[1])`.
3. Each decode step: `step_logits, step_weights = model.forward(generated[-1], cache=cache, start_pos=position, return_attention=record_attention)`, then `_absorb_attention(cache, step_weights, num_queries=1)`.
4. Gate the attention on policy type — never materialise attention weights when no policy consumes them, exactly as `perplexity` and `GenerationEngine` already do.

A regression test should assert that an `uses_attention=True` policy's `cum_attention` is non-zero during a needle evaluation, otherwise this silently regresses again.
